api/nakadi-event-bus-api.yaml

```diff
@@ -623,6 +623,50 @@ paths:
           schema:
             $ref: '#/definitions/Problem'
 
+  '/event-types/{name}/cursors':
+      get:
+        tags:
+          - unmanaged-api
+          - monitoring
+          - management-api
+        security:
+          - oauth2: ['nakadi.event_stream.read']
+        description: |
+          Transforms a list of Cursors with `shift` into a list without `shift`s. This is useful when there is the need
+          for randomly access events in the stream.
+
+        parameters:
+          - name: name
+            in: path
+            description: EventType name
+            type: string
+            required: true
+          - name: X-Flow-Id
+            in: header
+            description: |
+              The flow id of the request, which is written into the logs and passed to called services. Helpful
+              for operational troubleshooting and log analysis.
+            type: string
+          - name: X-nakadi-cursors
+            in: header
```

Comment:
I'll try to concentrate on this specific API here instead of the general guideline issue.

The API in the current form takes a list of "shifted cursor" objects, where each is a combination of partition, offset and (optional) shift, and produces a list of cursor objects (partition/offset).
(I guess the idea is that each input array entry corresponds to one result array entry, in the same order?.)

One problem here is that OpenAPI in its 2.0 incarnation doesn't allow putting complex objects (or arrays thereof) into a header, which manifests in using the reference to the schema in a "format" property, because there is no proper place for that. It is not clear to me how this array of objects would be encoded in the header.

The other issue is that the _main input_ of an operation, not just a context information like language, accepted media type, request content type etc., is hidden in a header, instead of being expressed in a URL (or possibly a POST body).

I would suggest putting the three values – partition, offset and shift – into query parameters. Of course, this makes it complicated to have a bulk request as it is currently, but is this bulk request really needed? If so, you could possibly put the list of objects into a POST body.